archive_verify/pdc_client.py

```diff
@@ -67,46 +92,42 @@ def cleanup(self):
     @staticmethod
     def _parse_dsmc_return_code(exit_code, output, whitelist):
         """
-        Parses the dsmc output when we've encountered a non-zero exit code. For some certain exit codes,
-        warnings and errors we still want to return successfully.
+        Parses the dsmc output when we've encountered a non-zero exit code. For some certain exit
+        codes, warnings and errors we still want to return successfully.
 
         :param exit_code: The exit code received from the failing dsmc process
         :param output: The text output from the dsmc process
         :param whitelist: A list of whitelisted warnings
         :returns True if only whitelisted warnings was encountered in the output, otherwise False
         """
-        log.info("DSMC process returned an error!")
 
         # DSMC sets return code to 8 when a warning was encountered.
-        if exit_code == 8:
-            log.info("DSMC process actually returned a warning.")
+        log_fn = log.warning if exit_code == 8 else log.error
+        log_fn(f"DSMC process returned a{' warning' if exit_code == 8 else 'n error'}!")
 
-            output = output.splitlines()
+        # parse the DSMC output and extract error/warning codes and messages
+        codes = []
+        for line in output.splitlines():
+            if line.startswith("ANS"):
+                log_fn(line)
 
-            # Search through the DSMC log and see if we only have
-            # whitelisted warnings. If that is the case, change the
-            # return code to 0 instead. Otherwise keep the error state.
-            warnings = []
+            matches = re.findall(r'ANS[0-9]+[EW]', line)
+            for match in matches:
+                codes.append(match)
 
-            for line in output:
-                matches = re.findall(r'ANS[0-9]+W', line)
+        unique_codes = set(sorted(codes))
+        if unique_codes:
+            log_fn(f"ANS codes found in DSMC output: {', '.join(unique_codes)}")
 
-                for match in matches:
-                    warnings.append(match)
+            # if we only have whitelisted warnings, change the return code to 0 instead
+            if not unique_codes.difference(set(whitelist)):
```

Comment:
To me it would perhaps be a bit more intuitive formulate it as:
```suggestion
            if unique_codes.issubset(set(whitelist)):
```

Reply:
nice, agreed, thanks!